bonsai-core/src/main/scala/com/stripe/bonsai/layout/DenseLayout.scala

```diff
@@ -5,20 +5,266 @@ import scala.language.higherKinds
+        val arr = classTag[A].newArray(length)
+        readArray(in, arr)
+        cons(arr)
+
+      case DenseArrayLayout.ByteDictionaryEncoding =>
+        val dictLen = in.readInt()
+        require(dictLen <= 255)
+        val dict: Array[A] = new Array[A](dictLen)
+        readArray(in, dict)
+        val encoding: Array[Byte] = new Array[Byte](in.readInt())
+        DenseArrayLayout.readByte(in, encoding)
+        // Ideally we would just use the dict encoding directly, but
+        // this will suffice for now.
+        val arr: Array[A] = new Array[A](encoding.length)
+        var i = 0
+        while (i < arr.length) {
+          arr(i) = dict(encoding(i).toInt & 0xFF)
+          i += 1
+        }
+        cons(arr)
+
+      case _ =>
+        throw new java.io.IOException("unsupported encoding for dense array layout")
+    }
+  }
+
+  def write(vec: Vec[A], out: DataOutput): Unit = {
+    val arr: Array[A] = vec match {
+      case (vec: DenseVec[_]) => vec.vec
+      case _ => vec.toArray
+    }
+
+    dictionaryFor(arr) match {
+      case Some((dict, encoding)) =>
+        out.writeByte(DenseArrayLayout.ByteDictionaryEncoding)
+        out.writeInt(dict.length)
+        writeArray(dict, out)
+        out.writeInt(encoding.length)
+        DenseArrayLayout.writeByte(encoding, out)
+
+      case None =>
+        out.writeByte(DenseArrayLayout.PlainEncoding)
+        out.writeInt(arr.length)
+        writeArray(arr, out)
+    }
+  }
+
+  private def dictionaryFor(arr: Array[A]): Option[(Array[A], Array[Byte])] = {
+    import scala.collection.mutable.HashMap
+    val dict: HashMap[A, Int] = new HashMap
+    val encoding: Array[Byte] = new Array[Byte](arr.length)
+    var i: Int = 0
+    while (i < arr.length) {
+      val a = arr(i)
+      val key = dict.get(a) match {
+        case Some(k) =>
+          k
+        case None =>
+          val k = dict.size
+          if (k > 255)
+            return None
+          dict.put(a, k)
+          k
+      }
+      encoding(i) = key.toByte
+      i += 1
+    }
+    val flatDict: Array[A] = new Array[A](dict.size)
+    dict.foreach { case (a, k) =>
+      flatDict(k) = a
+    }
+    Some((flatDict, encoding))
+  }
 }
 
-case class DenseArrayLayout[A: ClassTag]() extends Layout[A] {
-  def newBuilder =
-    new DenseBuilder[A, Array[A]](ArrayBuilder.make[A](), DenseArrayVec(_))
+object DenseArrayLayout {
+  val PlainEncoding = 0.toByte
+  val ByteDictionaryEncoding = 1.toByte
+
+  def readBoolean(in: DataInput, arr: Array[Boolean]): Unit = {
+    var i = 0
+    while (i < arr.length) {
+      arr(i) = in.readBoolean()
+      i += 1
+    }
+  }
+
+  def readByte(in: DataInput, arr: Array[Byte]): Unit = {
+    var i = 0
+    while (i < arr.length) {
+      arr(i) = in.readByte()
+      i += 1
+    }
+  }
+
+  def readShort(in: DataInput, arr: Array[Short]): Unit = {
+    var i = 0
+    while (i < arr.length) {
+      arr(i) = in.readShort()
+      i += 1
+    }
+  }
```

Comment:
Not really something you have to change, but it would be really nice if we had something like `cfor` (or equivalent) for all the while loops in this PR.